File: addons/guide-cpp/src/modifiers/guide_modifier_input_swizzle.cpp
#include "guide_modifier_input_swizzle.h"

using namespace godot;

void GUIDEModifierInputSwizzle::_bind_methods() {
    BIND_ENUM_CONSTANT(YXZ);
    BIND_ENUM_CONSTANT(ZYX);
    BIND_ENUM_CONSTANT(XZY);
    BIND_ENUM_CONSTANT(YZX);
    BIND_ENUM_CONSTANT(ZXY);

    ClassDB::bind_method(D_METHOD("get_order"), &GUIDEModifierInputSwizzle::get_order);
    ClassDB::bind_method(D_METHOD("set_order", "order"), &GUIDEModifierInputSwizzle::set_order);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "order", PROPERTY_HINT_ENUM, "YXZ,ZYX,XZY,YZX,ZXY"), "set_order", "get_order");
}

GUIDEModifierInputSwizzle::GUIDEModifierInputSwizzle() {
}

GUIDEModifierInputSwizzle::~GUIDEModifierInputSwizzle() {
}

Vector3 GUIDEModifierInputSwizzle::_modify_input(Vector3 input, double delta, int value_type) const {
    switch (order) {
        case YXZ: return Vector3(input.y, input.x, input.z);
        case ZYX: return Vector3(input.z, input.y, input.x);
        case XZY: return Vector3(input.x, input.z, input.y);
        case YZX: return Vector3(input.y, input.z, input.x);
        case ZXY: return Vector3(input.z, input.x, input.y);
    }
    return input;
}

bool GUIDEModifierInputSwizzle::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifierInputSwizzle> o = other;
    if (o.is_null()) return false;
    return o->get_order() == (int)order;
}

String GUIDEModifierInputSwizzle::_editor_name() const {
    return "Input Swizzle";
}

String GUIDEModifierInputSwizzle::_editor_description() const {
    return "Swizzle the input vector components. Useful to map 1D input to 2D or vice versa.";
}

File: addons/GuideCS/Modifiers/GUIDEModifierInputSwizzle.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierInputSwizzle : GUIDEModifier {
	public enum GUIDEInputSwizzleOperation {
		YXZ,
		ZYX,
		XZY,
		YZX,
		ZXY
	}

	[Export] public GUIDEInputSwizzleOperation Order { get; set; } = GUIDEInputSwizzleOperation.YXZ;

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifierInputSwizzle otherSwizzle && Order == otherSwizzle.Order;
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		switch (Order) {
			case GUIDEInputSwizzleOperation.YXZ:
				return new Vector3(input.Y, input.X, input.Z);
			case GUIDEInputSwizzleOperation.ZYX:
				return new Vector3(input.Z, input.Y, input.X);
			case GUIDEInputSwizzleOperation.XZY:
				return new Vector3(input.X, input.Z, input.Y);
			case GUIDEInputSwizzleOperation.YZX:
				return new Vector3(input.Y, input.Z, input.X);
			case GUIDEInputSwizzleOperation.ZXY:
				return new Vector3(input.Z, input.X, input.Y);
			default:
				GD.PushError($"Unknown order {Order} this is most likely a bug, please report it.");
				return input;
		}
	}

	public override string _EditorName() => "Input Swizzle";

	public override string _EditorDescription() => "Swizzle the input vector components. Useful to map 1D input to 2D or vice versa.";
}
